Regression (2.9.0): symlinked `node_modules/.bin` entry not realpath'd when a `package.json` is present

Version: Deno 2.9.0

**Broken:** `deno 2.9.0` · **Works:** `deno 2.8.3` (stable, x86_64-unknown-linux-gnu, Linux)

With a `package.json` present (node-compat mode), 2.9.0 uses a **symlinked** entry's own path as the main-module URL instead of realpath-ing it, so the entry's relative `import`s resolve one directory wrong. 2.8.3, Node, and Bun realpath the entry first. This breaks `node_modules/.bin/<tool>` shims that import siblings via relative paths.

### Repro (no npm)
```bash
mkdir -p proj/node_modules/.bin proj/node_modules/pkg/bin proj/node_modules/pkg/lib
echo '{}'                                                > proj/package.json
echo '{"name":"pkg","version":"1.0.0","type":"module"}'  > proj/node_modules/pkg/package.json
printf "import {x} from '../lib/x.js';\nconsole.log('loaded ok:', x);\n" > proj/node_modules/pkg/bin/cli.js
echo "export const x = 'OK';"                            > proj/node_modules/pkg/lib/x.js
ln -s ../pkg/bin/cli.js proj/node_modules/.bin/cli       # as npm creates .bin entries
cd proj && deno run --allow-all --ext=js node_modules/.bin/cli
```
(`.bin/cli` is extensionless like real npm shims, so `--ext=js`; same failure via the shebang `#!/usr/bin/env -S deno run --allow-all --ext=js`.)

**2.9.0:**
```
error: [ERR_MODULE_NOT_FOUND] Cannot find module 'file:///…/proj/node_modules/lib/x.js' imported from 'file:///…/proj/node_modules/.bin/cli'
```
`../lib/x.js` resolved against the symlink's dir (`node_modules/.bin/`) instead of the target's (`node_modules/pkg/bin/`).
**2.8.3 / Node / Bun:** `loaded ok: OK`

### The trigger is the `package.json`
`rm proj/package.json` → 2.9.0 prints `loaded ok: OK` (plain resolution still realpaths). Running the resolved target `node_modules/pkg/bin/cli.js` also works on 2.9.0. So the regression is specific to symlinked entries resolved through the `node_modules`/package.json path.

### Re: #35545 / #35547
That fixture is a `.bin/cli` **not under `node_modules` and with no `package.json`** — plain file resolution, which realpaths on every version, so it never hits the regressed path and looks like "not a regression." Under `node_modules` **with** a `package.json` it reproduces: 2.8.3 ✓ → 2.9.0 ✗.

--- a/cli/module_loader.rs
+++ b/cli/module_loader.rs
@@ -84,6 +84,7 @@ use node_resolver::InNpmPackageChecker;
 use node_resolver::NodeResolutionKind;
 use node_resolver::ResolutionMode;
 use node_resolver::errors::PackageJsonLoadError;
+use sys_traits::FsCanonicalize;
 use sys_traits::FsMetadata;
 use sys_traits::FsMetadataValue;
 use sys_traits::FsRead;
@@ -1147,15 +1148,20 @@ impl<TGraphContainer: ModuleGraphContainer>
     // `file:///<project>/` mapping whose prefix match rewrites any file URL
     // outside the project root (such as the npm cache) to live underneath the
     // project directory, breaking the load. Such a package-internal main
-    // module is not a user source file and must not be remapped, so load it
-    // verbatim. A user-provided path entry (e.g. `deno run ./thing.ts`) is
+    // module is not a user source file and must not be remapped, so bypass the
+    // import map. A user-provided path entry (e.g. `deno run ./thing.ts`) is
     // intentionally left to the import map.
     if matches!(kind, deno_core::ResolutionKind::MainModule)
       && let Ok(url) = ModuleSpecifier::parse(raw_specifier)
       && url.scheme() == "file"
       && self.shared.in_npm_pkg_checker.in_npm_package(&url)
     {
-      return Ok(url);
+      let canonicalized_url = deno_path_util::url_to_file_path(&url)
+        .ok()
+        .and_then(|path| self.shared.sys.fs_canonicalize(&path).ok())
+        .and_then(|path| deno_path_util::url_from_file_path(&path).ok())
+        .unwrap_or(url);
+      return Ok(canonicalized_url);
     }
 
     let referrer = self

--- a/tests/specs/npm/run_bin_symlink_realpath_package_json/__test__.jsonc
+++ b/tests/specs/npm/run_bin_symlink_realpath_package_json/__test__.jsonc
@@ -1,0 +1,14 @@
+{
+  "tempDir": true,
+  "if": "unix",
+  "steps": [
+    {
+      "args": "run -A setup.js",
+      "output": "[WILDCARD]"
+    },
+    {
+      "args": "run -A --ext=js node_modules/.bin/cli",
+      "output": "loaded ok: OK\n"
+    }
+  ]
+}

--- a/tests/specs/npm/run_bin_symlink_realpath_package_json/setup.js
+++ b/tests/specs/npm/run_bin_symlink_realpath_package_json/setup.js
@@ -1,0 +1,18 @@
+Deno.mkdirSync("node_modules/.bin", { recursive: true });
+Deno.mkdirSync("node_modules/pkg/bin", { recursive: true });
+Deno.mkdirSync("node_modules/pkg/lib", { recursive: true });
+
+Deno.writeTextFileSync("package.json", "{}\n");
+Deno.writeTextFileSync(
+  "node_modules/pkg/package.json",
+  '{"name":"pkg","version":"1.0.0","type":"module"}\n',
+);
+Deno.writeTextFileSync(
+  "node_modules/pkg/bin/cli.js",
+  "import { x } from '../lib/x.js';\nconsole.log('loaded ok:', x);\n",
+);
+Deno.writeTextFileSync(
+  "node_modules/pkg/lib/x.js",
+  "export const x = 'OK';\n",
+);
+Deno.symlinkSync("../pkg/bin/cli.js", "node_modules/.bin/cli");

